numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -1028,7 +1094,28 @@ PyArray_IntpFromIndexSequence(PyObject *seq, npy_intp *vals, npy_intp maxvals)
 NPY_NO_EXPORT int
 PyArray_IntpFromSequence(PyObject *seq, npy_intp *vals, int maxvals)
 {
-    return PyArray_IntpFromIndexSequence(seq, vals, (npy_intp)maxvals);
+    if (PyLong_CheckExact(seq) || !PySequence_Check(seq)) {
+        npy_intp rvalue = intp_from_scalar(seq, vals, 0);
+
+        /*
+        * Check if an error occurred.
+        */
+        if (PyErr_Occurred() != NULL || rvalue != vals[0]) {
+            return -1;
+        }
+
+        return 1;
+    }
+
+    PyObject *seq_obj = PySequence_Fast(seq,
+        "An error occurred during a call to `PySequence_Fast`, an iterable "
+        "object is required. The given object is not a scalar integer nor a "
+        "sequence.");
```

Comment:
Same here, this is fine, but the error message is not helpful.  Maybe the default is even OK (you could compare what it was before and after for an invalid object?).

Reply:
Fixed in 12c9b4e.

I was not able to recover the default error message since `PySequence_Fast()` strongly requires the message, which means that passing `NULL` results in a segmentation fault, and passing `""` results in `numpy/core/tests/test_multiarray.py:526: TypeError`, which does not contain any error message.

I replaced my message with yours (slightly modified).